executor/index_merge_reader.go

```diff
@@ -0,0 +1,738 @@
+	startTS    uint64
+	// handleIdx is the index of handle, which is only used for case of keeping order.
+	handleIdx    int
+	tableRequest *tipb.DAGRequest
+	// columns are only required by union scan.
+	columns           []*model.ColumnInfo
+	partialStreamings []bool
+	tableStreaming    bool
+	*dataReaderBuilder
+	// All fields above are immutable.
+
+	partialWorkerWg sync.WaitGroup
+	tblWorkerWg     sync.WaitGroup
+	processWokerWg  sync.WaitGroup
+	finished        chan struct{}
+
+	kvRanges      []kv.KeyRange
+	workerStarted bool
+
+	resultCh   chan *lookupTableTask
+	resultCurr *lookupTableTask
+	feedbacks  []*statistics.QueryFeedback
+
+	// memTracker is used to track the memory usage of this executor.
+	memTracker *memory.Tracker
+
+	// checkIndexValue is used to check the consistency of the index data.
+	*checkIndexValue
+
+	corColInIdxSide bool
+	partialPlans    [][]plannercore.PhysicalPlan
+	corColInTblSide bool
+	tblPlans        []plannercore.PhysicalPlan
+	corColInAccess  bool
+	idxCols         [][]*expression.Column
+	colLens         [][]int
+}
+
+// Open implements the Executor Open interface
+func (e *IndexMergeReaderExecutor) Open(ctx context.Context) error {
+	kvRangess := make([][]kv.KeyRange, 0, len(e.partialPlans))
+	for i := 0; i < len(e.partialPlans); i++ {
+		_, ok := e.partialPlans[i][0].(*plannercore.PhysicalIndexScan)
+		if ok {
+			kvRanges, err := distsql.IndexRangesToKVRanges(e.ctx.GetSessionVars().StmtCtx, getPhysicalTableID(e.table), e.indexes[i].ID, e.ranges[i], e.feedbacks[i])
+			if err != nil {
+				return errors.Trace(err)
+			}
+			kvRangess = append(kvRangess, kvRanges)
+		} else {
+			kvRangess = append(kvRangess, nil)
+		}
+	}
+	err := e.open(ctx, kvRangess)
+	return errors.Trace(err)
+}
+
+func (e *IndexMergeReaderExecutor) open(ctx context.Context, kvRangess [][]kv.KeyRange) error {
+	e.finished = make(chan struct{})
+	e.resultCh = make(chan *lookupTableTask, atomic.LoadInt32(&LookupTableTaskChannelSize))
+
+	workCh := make(chan *lookupTableTask, 1)
+	fetchCh := make(chan *lookupTableTask, len(kvRangess))
+
+	go e.startIndexMergeProcessWorker(ctx, workCh, fetchCh, len(kvRangess))
+	e.partialWorkerWg.Add(len(kvRangess))
+	for i := 0; i < len(kvRangess); i++ {
+		if e.indexes[i] != nil {
+			err := e.startPartialIndexWorker(ctx, kvRangess[i], fetchCh, i)
+			if err != nil {
+				return err
+			}
+		} else {
+			e.startPartialTableWorker(ctx, fetchCh, i)
+		}
+	}
+	go e.startIndexMergeTableScanWorker(ctx, workCh)
+	e.workerStarted = true
+	return nil
+}
+
+func (e *IndexMergeReaderExecutor) startIndexMergeProcessWorker(ctx context.Context, workCh chan<- *lookupTableTask, fetch <-chan *lookupTableTask, partialWorkerCount int) {
+	idxMergeProcessWorker := &indexMergeProcessWorker{
+		fetchCh:  fetch,
+		workCh:   workCh,
+		resultCh: e.resultCh,
+		maps:     make(map[int64]byte),
+	}
+	e.processWokerWg.Add(1)
+	go func() {
+		idxMergeProcessWorker.fetchLoop(ctx, partialWorkerCount)
+		e.processWokerWg.Done()
```

Comment:
How about putting this statement into `idxMergeProcessWorker.fetchLoop(ctx, partialWorkerCount)` to make here more simple like `go idxMergeProcessWorker.fetchLoop(ctx, partialWorkerCount)`.

Reply:
`e.processWokerWg.Done()` should be after `idxMergeProcessWorker.fetchLoop`. If simplifying, `Done()` may be with a wrong behavior.